### PR TITLE
Shift Management

### PR DESCRIPTION
**Pull request by PawanMeh**
1. Shift Request

shift_assignment

2. Shift Assignment

shift_calendar

### CODE REVIEW

File: erpnext/hr/doctype/shift_request/shift_request.py [hunk 1]
```diff
@@ -4,7 +4,83 @@
...
+				where employee = %(employee)s and docstatus < 2
+				and to_date >= %(from_date)s and from_date <= %(to_date)s
+				and name != %(name)s""", {
+					"employee": self.employee,
+					"from_date": self.from_date,
+					"to_date": self.to_date,
+					"name": self.name
+				}, as_dict = 1):
+
+				if (((getdate(self.from_date) >= getdate(d.to_date)
```
Comment (nabinhait) [inline erpnext/hr/doctype/shift_request/shift_request.py]: Why checking in python? You can write the conditions in query itself.

File: erpnext/hr/doctype/shift_request/shift_request.py [hunk 2]
```diff
@@ -4,7 +4,83 @@
...
+					self.shift_type == d.shift_type ):
+						self.throw_overlap_error(d)
+
+	def throw_overlap_error(self, d):
+		msg = _("Employee {0} has already applied for {1} between {2} and {3} : ").format(self.employee,
+			d['shift_type'], formatdate(d['from_date']), formatdate(d['to_date'])) \
+			+ """ <b><a href="#Form/Shift Request/{0}">{0}</a></b>""".format(d["name"])
+		frappe.throw(msg, OverlapError)
+
+	def get_date_list(self, start_date, end_date):
```
Comment (nabinhait) [inline erpnext/hr/doctype/shift_request/shift_request.py]: get_working_days ?

File: erpnext/hr/doctype/shift_assignment/shift_assignment.py [hunk 3]
```diff
@@ -4,7 +4,76 @@
...
 class ShiftAssignment(Document):
-	pass
+	def validate(self):
+		self.validate_overlapping_dates();
+
+	def validate_overlapping_dates(self):
+			if not self.name:
+				self.name = "New Shift Assignment"
+
+			for d in frappe.db.sql("""
```
Comment (nabinhait) [inline erpnext/hr/doctype/shift_assignment/shift_assignment.py]: no need for a for loop, no possibility for multiple. Add date condition in query itself.

File: erpnext/hr/doctype/shift_assignment/shift_assignment.py [hunk 4]
```diff
@@ -4,7 +4,76 @@
...
+					"date": self.date,
+					"name": self.name
+				}, as_dict = 1):
+
+				if (getdate(self.date) == getdate(d.date)
+					and self.shift_type == d.shift_type ):
+						self.throw_overlap_error(d)
+
+	def throw_overlap_error(self, d):
+		msg = _("Employee {0} has already applied for {1} between {2} and {3} : ").format(self.employee,
```
Comment (nabinhait) [inline erpnext/hr/doctype/shift_assignment/shift_assignment.py]: `between {2} and {3}` ???
there is only one date.

**Review by nabinhait — CHANGES_REQUESTED** [review]

File: erpnext/hr/doctype/shift_request/shift_request.py (hunk 2 above)
Comment (PawanMeh) [reply erpnext/hr/doctype/shift_request/shift_request.py]: @nabinhait , made the other review comment changes. Related to this, may need some more clarification.

Comment (nabinhait) [reply erpnext/hr/doctype/shift_request/shift_request.py]: Renaming the function to `get_working_dates` will clear the objective of the function.

**Comment by PawanMeh** [comment]
Made requested changes

**Pull request merged**